catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java

```diff
@@ -127,31 +119,18 @@ public void uploadCatalog(final String catalogXML, final CallContext callContext
             catalogCache.clearCatalog(internalTenantContext);
         } catch (final TenantApiException e) {
             throw new CatalogApiException(e);
-        } catch (final ValidationException e) {
-            throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, internalTenantContext.getTenantRecordId());
-        } catch (final JAXBException e) {
-            throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, internalTenantContext.getTenantRecordId());
-        } catch (final IOException e) {
-            throw new IllegalStateException(e);
-        } catch (final TransformerException e) {
-            throw new IllegalStateException(e);
-        } catch (final SAXException e) {
-            throw new IllegalStateException(e);
         }
     }
 
     @Override
-    public void validateCatalog(final String catalogXML, final CallContext context) throws CatalogApiException {
+    public CatalogValidation validateCatalog(final String catalogXML, final CallContext context) {
         final InternalTenantContext internalTenantContext = createInternalTenantContext(context);
-        try {
-            XMLLoader.getObjectFromStream(new ByteArrayInputStream(catalogXML.getBytes(StandardCharsets.UTF_8)), StandaloneCatalog.class);
-        } catch (final ValidationException e) {
-            throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, internalTenantContext.getTenantRecordId());
-        } catch (final JAXBException e) {
-            throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, internalTenantContext.getTenantRecordId());
-        } catch (final Exception e) {
-            throw new RuntimeException(e);
+        final ValidationErrors errors = validateCatalogInternal(catalogXML, internalTenantContext);
+        if (!errors.isEmpty()) {
+            logger.info("Failed to load new catalog version: " + errors.toString());
         }
+        catalogCache.clearCatalog(internalTenantContext);
```

Comment:
I don't think this is needed?

Reply:
This is required. The [validateCatalogInternal](https://github.com/killbill/killbill/blob/ce2ebe8b05df684f9d6830e81bed8b4791e19f6a/catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java#L206) method [adds](https://github.com/killbill/killbill/blob/ce2ebe8b05df684f9d6830e81bed8b4791e19f6a/catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java#L222) the new catalog to be validated to the existing `VersionedCatalog` in order to perform the validation. This results in the `VersionedCatalog` in the cache being modified during the validation process without an actual catalog upload being performed. So, the cache needs to be cleared after the validation is complete.

Reply:
oh, I missed this. This is risky...

Maybe we should clone the catalog first to be safe (using `readExternal`/`writeExternal`). @sbrossie thoughts?